src/spatial_data_format.rs

```diff
@@ -0,0 +1,515 @@
+    /// the delimiter between the header and sample binary.
+    NoDelimiter,
+
+    /// Returned when somehow we fail to turn four bytes into a f32 when reading.
+    TryInto,
+
+    /// Returned when io fails when reading or writing files.
+    IoError(std::io::Error),
+
+    /// Returned when serialization of the header fails.
+    RonError(ron::Error),
+
+    /// Returned when deserialization of the header fails.
+    RonSpannedError(ron::de::SpannedError),
+}
+
+impl fmt::Display for GrapeFileError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use GrapeFileError as GFE;
+        let msg = match self {
+            GFE::UnequalSampleBufferLengths => Cow::from("unequal sample buffer lengths"),
+            GFE::NoDelimiter => Cow::from("no delimiter in GrapeFile"),
+            GFE::TryInto => Cow::from("something went wrong while parsing f32s"),
+            GFE::IoError(error) => Cow::from(format!("io error: {}", error)),
+            GFE::RonError(error) => Cow::from(format!("ron error: {}", error)),
+            GFE::RonSpannedError(error) => Cow::from(format!("ron spanning error: {}", error)),
+        };
+
+        write!(f, "{}", msg)
+    }
+}
+
+impl std::error::Error for GrapeFileError {}
+
+impl GrapeFile {
+    /// Make a [GrapeFileBuilder], which can be used to set sample rate and
+    /// add streams, before building the [GrapeFile].
+    pub fn builder() -> GrapeFileBuilder {
+        GrapeFileBuilder::new()
+    }
+
+    /// Write out a [GrapeFile] to the path provided.
+    pub fn to_path(&self, path: impl AsRef<Path>) -> Result<(), GrapeFileError> {
+        let mut handle = File::create(path).map_err(GrapeFileError::IoError)?;
+        self.to_file(&mut handle)
+    }
+
+    /// Write out a [GrapeFile] to the [Write]able object provided.
+    pub fn to_file(&self, file: &mut impl Write) -> Result<(), GrapeFileError> {
+        let h_str = ron::ser::to_string(&self.header).map_err(GrapeFileError::RonError)?;
+        let h_buf = h_str.as_bytes();
+
+        file.write_all(h_buf).map_err(GrapeFileError::IoError)?;
+
+        file.write_all(&[0xFF]).map_err(GrapeFileError::IoError)?;
+
+        let s_buf: Vec<u8> = self.samples.iter().flat_map(|f| f.to_be_bytes()).collect();
+
+        file.write_all(&s_buf).map_err(GrapeFileError::IoError)?;
+
+        Ok(())
+    }
+
+    /// Read a [GrapeFile] from the path provided.
+    pub fn from_path(path: impl AsRef<Path>) -> Result<Self, GrapeFileError> {
+        let mut handle = File::open(path).map_err(GrapeFileError::IoError)?;
+        Self::from_file(&mut handle)
+    }
+
+    /// Read a [GrapeFile] from the [Read]able object provided.
+    pub fn from_file(file: &mut impl Read) -> Result<Self, GrapeFileError> {
+        let mut raw_text = Vec::new();
+        file.read_to_end(&mut raw_text)
+            .map_err(GrapeFileError::IoError)?;
+
+        let delim_idx = raw_text
+            .iter()
+            .position(|e| *e == 0xFF)
```

Comment:
Great point. Luckily `ron` serializes into human-readable data, like JSON (see [here](https://docs.rs/ron/latest/ron/index.html#ron-syntax-overview)). 255 isn't an ascii character (printable or otherwise) so we know it won't appear in the header.

I did just write a test that adds 255 streams though, just in case. It'll be pushed once I'm done working through these comments :)